### ISSUE TITLE
SetUp and TearDown for all fixtures in namespace

### ISSUE THREAD

**Issue by Kralizek** [nunit/nunit#3140]
Hi!

I find myself repeating the same teardown code on all test fixtures in a given namespace. I know I could solve this scenario with inheritance, but I think that there is a fit for extending the `SetUpFixture` class.

Ideally I would like to have a `[TearDown]` method in the SetUpFixture class and have that method being executed at teardown-time for all the tests in the namespace.

Here is an example that extends the sample availe in the documentation here: https://github.com/nunit/docs/wiki/SetUpFixture-Attribute

```csharp
using System;
using NUnit.Framework;

namespace NUnit.Tests
{
  [SetUpFixture]
  public class MySetUpClass
  {
	[OneTimeSetUp]
	public void RunBeforeAnyTests()
	{
	  // ...
	}

	[SetUp]
	public void RunBeforeEachTest()
	{
	  // ...
	}

	[TearDown]
	public void RunAfterEachTest()
	{
	  // ...
	}

	[OneTimeTearDown]
	public void RunAfterAllTests()
	{
	  // ...
	}
  }
}
```

Thanks!

**Comment by rprouse** [comment]
Interesting idea. I think it might be a confusing feature, but I can see your use case. What do other people think?

**Comment by Kralizek** [comment]
@rprouse thanks for your comment. For additional context, I can give you a concrete example where this feature would be very useful.

In this folder I have a set of test fixtures whose tests are creating some resources on a remote server. At the end of every test, I want those resources to be removed.

https://github.com/Nybus-project/Nybus/tree/master/tests/integration/Tests.Integration.Nybus.Engine.RabbitMq/External

**Comment by drewburlingame** [comment]
I like this requiest but I also think it could be confusing if you were new to the tests and had to debug them.  If it's not obvious from reading the test that it already has setup & teardown methods, you wouldn't know where to go looking for the setup methods if you even thought of looking for them.  It would remove the confusion if you referenced the class from a test fixture.   For example  
```
[UseSetupAndTeardownFrom(typeof(MySetUpClass)]
public class MyTestClass
```
or

```
public class MyTestClass : IUseSetupAndTeardownFrom<MySetUpClass>
```

What would be the behavior if you had to Setup methods defined?

**Comment by Kralizek** [comment]
@drewburlingame I see your point but isn't the whole SetUpFixture functionality already obscure by definition? After all, it's a class with an attribute dropped in a folder that affects all the fixtures of the folder without them even needing to know if exists.

Besides, not having access to the fields and methods of the fixture, it is going to execute teardown code that shouldn't be affecting them.

**Comment by drewburlingame** [comment]
@Kralizek you're right.  the SetUpFixture is obscure by definition.  I see the benefits of the ease of use when writing these tests initially when you have all the context.  You would need to determine how to handle cases where nested hierarchies declared the SetUpFixture.  The easy solution is to throw an exception.  This will force devs to move SetUpFixtures down N levels, which could also introduce duplication.  Another option is for the closest ancestor namespace to win, but I think that could create confusion for devs new to the project who somehow miss the closest ancestor and wonder why another one isn't working.  This can also be confusing for devs who step into a newer version of NUnit and don't know this feature has been added or how it behaves.

If you specify the setup class to use for a fixture, it has the benefits of being obvious to all developers and it also enables a more flexible composition.

**Comment by drewburlingame** [comment]
@Kralizek The comment on this post by Rob Prouse seems to indicate this is already a feature.  https://stackoverflow.com/a/44591287/169336

**Comment by CharliePoole** [comment]
An added complexity is that SetUp and Teardown methods in SetUpFixtures would have to work in a predictable fashion alongside SetUp and TearDown in case classes. The question of whether to run or skip TearDown in a given case becomes especially tricky. And if it's tricky to figure it out as implementors it's going to be harder for users.

**Comment by CharliePoole** [comment]
What's not an issue is how to handle multiple SetUpFixtures.

**Comment by Kralizek** [comment]
@drewburlingame I'll give it a try although the snippet attached right after seems to go not into that direction.

As for @charliepoole question, I'd assume that code placed in a SetUpFixture class gets always executed.

**Comment by CharliePoole** [comment]
@Kralizek Sorry, but that's an incorrect assumption about how SetUp and TearDown are defined to work.

Let's say that a given test, through inheritance and nesting, has a list of 4 pairs of SetUp and TearDown methods.  We can number them in order of execution, 1 through 4.

Then, normal execution is SetUp 1, 2, 3, 4, Test, TearDown 4, 3, 2, 1.

But suppose SetUp  2 throws an exception. Then we should see SetUp 1, 2, TearDown 2, 1.

Any new kindy of global SetUp you introduce has to fit into the same logic or it won't make sense.

**Comment by Kralizek** [comment]
@CharliePoole thank you for your explaination. I misunderstood your first post then. I was more looking into the perspective of "some tests might want or not want to execute a TearDown method placed in the SetUpFixture class", I wasn't thinking in terms of exceptions. Thanks again for your time explaining the issue.

**Comment by CharliePoole** [comment]
Ah, I see. No, hiaving a test "decide" whether to have its setup run would seem a bit odd to me and also hard to reconcile with some existing features.

**Comment by Kralizek** [comment]
> Ah, I see. No, hiaving a test "decide" whether to have its setup run would seem a bit odd to me and also hard to reconcile with some existing features.

Totally agree with you, hence my comment "I'd assume that code placed in a SetUpFixture class gets always executed."

**Comment by CharliePoole** [comment]
OneTimeSetUp code in a SetUpFixture is always executed unless a higher level SetUpFixture that wraps it threw an exception. IMO any other type of setup we add should follow the same rules.

My main point here was just to point out that there's more complexity here than initially meets the eye...

 * Multiple levels of TestFixture inheritance.
 * Multiple levels of SetUpFixture inheritance.
 * Nested namespaces with SetUpFixtures.
 * Multiple setups St the same level of inheritance are tolerated, but order of execution is non-deterministic.

Nothing is a deal-breaker but it all has to be dealt with. If I were starting over, it would be simpler. :smiley:

**Comment by baratgabor** [comment]
I was actually looking for this feature. It would be great for e.g. integration testing when the db has to be restored to a clean state after each test run; this is kind of a cross-cutting concern across the whole test assembly, so a centralized feature seems to be a better fit than deriving all test classes from a base class (which you can forget to do).
